File: python/robot-simulator/robot_simulator.py
EAST = 'east'
NORTH = 'north'
WEST = 'west'
SOUTH = 'south'


class Robot:
    def __init__(self, direction=NORTH, x_pos=0, y_pos=0):
        self.direction = direction
        self.coordinates = (x_pos, y_pos)

    def move(self, actions):
        for action in list(actions):
            match action:
                case 'L':
                    self.__left()
                case 'R':
                    self.__right()
                case 'A':
                    self.__advance()

    def __left(self):
        match self.direction:
            case 'north':
                self.direction = WEST
            case 'east':
                self.direction = NORTH
            case 'south':
                self.direction = EAST
            case 'west':
                self.direction = SOUTH

    def __right(self):
        match self.direction:
            case 'north':
                self.direction = EAST
            case 'east':
                self.direction = SOUTH
            case 'south':
                self.direction = WEST
            case 'west':
                self.direction = NORTH

    def __advance(self):
        match self.direction:
            case 'north':
                self.coordinates = (self.coordinates[0], self.coordinates[1] + 1)
            case 'east':
                self.coordinates = (self.coordinates[0] + 1 , self.coordinates[1])
            case 'south':
                self.coordinates = (self.coordinates[0], self.coordinates[1] - 1)
            case 'west':
                self.coordinates = (self.coordinates[0] - 1, self.coordinates[1])

File: javascript/robot-simulator/robot-simulator.js
export class InvalidInputError extends Error {
  constructor(message) {
    super()
    this.message = message || 'Invalid Input'
  }
}

export class Robot {
  constructor() {
    this.direction = 'north'
    this.x = 0
    this.y = 0
  }
  
  get bearing() {
    return this.direction
  }

  get coordinates() {
    return [this.x, this.y]
  }

  place({ x, y, direction }) {
    if (!['north', 'south', 'west', 'east'].includes(direction)) {
      throw new InvalidInputError('Invalid direction')
    }

    this.direction = direction
    this.x = x
    this.y = y
  }

  evaluate(instructions) {
    for (const instruction of instructions.split('')) {
      switch(instruction) {
        case 'L':
          this.left()
          break
          
        case 'R':
          this.right()
          break
          
        case 'A':
          this.advance()
          break
          
        default:
          throw new InvalidInputError('Not supported instruction')
      }
    }
  }

  left() {
    switch(this.bearing) {
      case 'north':
        this.direction = 'west'
        break
        
      case 'east':
        this.direction = 'north'
        break
        
      case 'south':
        this.direction = 'east'
        break
        
      case 'west':
        this.direction = 'south'
        break
    }
  }

  right() {
    switch(this.bearing) {
      case 'north':
        this.direction = 'east'
        break
        
      case 'east':
        this.direction = 'south'
        break
        
      case 'south':
        this.direction = 'west'
        break
        
      case 'west':
        this.direction = 'north'
        break
    }
  }

  advance() {
    switch(this.bearing) {
      case 'north':
        this.y++
        break
        
      case 'east':
        this.x++
        break
        
      case 'south':
        this.y--
        
        break
      case 'west':
        this.x--
        break
    }
  }
}
